### ISSUE TITLE
Volumes not initially mounting for services that restart on windows reboot

### ISSUE THREAD

**Issue by SpikyClip** [docker/for-win#13985]
### Description

I am having the exact same issue as #584 on Docker Desktop 4.28.0 (139021) for Windows.

Essentially on computer boot, when the service restarts, the containers cannot find the volumes so it appears like its the first setup. But when the services are restarted manually, the volumes are mounted with no issues.

This make startup services useless as they can't access the volumes they need. I suspect it has something to do with delayed WSL2 startup, or the mount paths I used in the compose file.

[docker-compose_truncated.yml.txt](https://github.com/docker/for-win/files/14720042/docker-compose_truncated.yml.txt)

### Reproduce

`sudo docker-compose up -d`

### Expected behavior

Volumes should be mounted correctly when docker desktop starts on windows reboot.

### docker version

```bash
Client:
 Cloud integration: v1.0.35+desktop.11
 Version:           25.0.3
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        4debf41
 Built:             Tue Feb  6 21:13:00 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Desktop
 Engine:
  Version:          25.0.3
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       f417435
  Built:            Tue Feb  6 21:14:25 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    25.0.3
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1-desktop.4
    Path:     /usr/local/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.6-desktop.1
    Path:     /usr/local/lib/docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.24
    Path:     /usr/local/lib/docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /usr/local/lib/docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.22
    Path:     /usr/local/lib/docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     /usr/local/lib/docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.0.1
    Path:     /usr/local/lib/docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /usr/local/lib/docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.5.0
    Path:     /usr/local/lib/docker/cli-plugins/docker-scout
WARNING: Plugin "/usr/local/lib/docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /usr/local/lib/docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 7
  Running: 6
  Paused: 0
  Stopped: 1
 Images: 7
 Server Version: 25.0.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
 Kernel Version: 5.15.146.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 7.734GiB
 Name: docker-desktop
 ID: a81fab99-f491-41fa-8fc1-09bc414d91f9
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: daemon is not using the default seccomp profile
```

### Diagnostics ID

E82A5D0F-B3D8-4BED-9935-9D986C6BB0D8/20240322090128

**Comment by Tristan-MyAnaPro** [comment]
@SpikyClip / @bsousaa Possibly a duplicate of the issue I encounter too ? https://github.com/docker/desktop-release/issues/59

**Comment by kasprk** [comment]
I'm experiencing the same issue. I have mariadb instance which mounts a local folder to /var/lib/mysql ... when I boot the computer, it does not mount the folder correctly although docker inspect shows it as mounted. But when you docker-compose exec into the container, I can see the default directory structure for freshly launched mariadb instance.
After restarting the container, the folder is mounted correctly.

**Comment by yolopunk** [comment]
I'm having the same problem with docker-desktop 4.28.0, the volume mount fails on a windows reboot, but it works fine on a manual startup. When I roll back the version of docker-desktop to 4.20.0, I don't have this problem.

**Comment by SpikyClip** [comment]
Unfortunately I am having difficulty downloading the installer for the [old version](https://docs.docker.com/desktop/release-notes/#4200):
![image](https://github.com/docker/for-win/assets/74215773/cfe20431-2f23-4018-93a8-8da7887f95ce)

**Comment by SpikyClip** [comment]
I tried downgrading to [`4.24.2`](https://docs.docker.com/desktop/release-notes/#4242) and and it works. So the bug is between that version and `4.28`. Don't really feel like testing all the version inbetween, but feel free to chime in if any later version works.

**Comment by barbuslex** [comment]
I have same issue in 4.30.0 :S

4.24.2 is dated of 12-10-2023... need fix please...

**Comment by nikhil96widhani** [comment]
Same issue for me, i downgraded for now but hope its fixed in next version

**Comment by longzheng** [comment]
I'm running into this bug on Windows as well with the [Powerwall-Dashboard](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/powerwall.yml) container (although it doesn't seem to affect my other [Teslamate](https://docs.teslamate.org/docs/installation/docker) container).

> I tried downgrading to [`4.24.2`](https://docs.docker.com/desktop/release-notes/#4242) and and it works. So the bug is between that version and `4.28`. Don't really feel like testing all the version inbetween, but feel free to chime in if any later version works.

I can also verify that `4.24.2` is working correctly.

I tested different versions all the way up to [`4.27.2`](https://docs.docker.com/desktop/release-notes/#4272) and it's still working so I think it's literally `4.28.0` that broke it.

**Comment by longzheng** [comment]
I can confirm that version `4.31.0` still has the same bug/issue.

**Comment by longzheng** [comment]
Looking at https://github.com/dockur/dnsmasq/issues/4, I also checked my `dockerd` [daemon logs](https://docs.docker.com/config/daemon/logs/) at `%LOCALAPPDATA%\Docker\log\vm\dockerd.log` and confirmed it's the same issue with bind mounts

```
time="2024-06-07T06:45:38.687149000Z" level=error msg="failed to start container" container=c47bbcc2866be37edb367e957c29f2c761c813d853e0119d76baf07a9c4a0377 error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/f8bdc6a9944521e050c4ef3a1e2ae3b2b39eb573242081361afb80a0c5a649e7\" to rootfs at \"/etc/telegraf/telegraf.conf\": stat /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/f8bdc6a9944521e050c4ef3a1e2ae3b2b39eb573242081361afb80a0c5a649e7: no such file or directory: unknown"
time="2024-06-07T06:45:38.871371700Z" level=error msg="failed to start container" container=843603cdcb15df27d3a8d374da931aa905fdfe695a80ee0686f2f86538460b79 error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/c964d0391d417d8719c015cbb73a53a85e64860f8fd33daa854b4eb49de7fd52\" to rootfs at \"/var/lib/grafana\": stat /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/c964d0391d417d8719c015cbb73a53a85e64860f8fd33daa854b4eb49de7fd52: no such file or directory: unknown"
time="2024-06-07T06:45:39.000972200Z" level=error msg="failed to start container" container=f89a22df31829d4870f8a0396148a0b1227e1b8431f05e8476128c58ee84ffb0 error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/a6326bcc0b9c72f750ab2a5f664b17f7c9a14530afa144b290774d0123ab3d6f\" to rootfs at \"/etc/influxdb/influxdb.conf\": stat /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/a6326bcc0b9c72f750ab2a5f664b17f7c9a14530afa144b290774d0123ab3d6f: no such file or directory: unknown"
time="2024-06-07T06:45:39.151204500Z" level=error msg="failed to start container" container=7d70b931b3924b604c1855243bcae5d2ffbd14647b84a223787a29b8edd7eb4a error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/2a2e53533e8c38a8fd40914c28b7fb037cb8f187499a3ff184eb9cc384acb143\" to rootfs at \"/app/.auth\": stat /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/2a2e53533e8c38a8fd40914c28b7fb037cb8f187499a3ff184eb9cc384acb143: no such file or directory: unknown"
time="2024-06-07T06:45:39.291634600Z" level=error msg="failed to start container" container=87ba2bebdd9d755e89d7741fc4ff271ef6d088418574c5cef267b3e423fa5cad error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/2af5984fa649aa521d7080a30d0e3da96686aa85c0ec42566802c195d1177cf4\" to rootfs at \"/var/lib/weather\": stat /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/2af5984fa649aa521d7080a30d0e3da96686aa85c0ec42566802c195d1177cf4: no such file or directory: unknown"
time="2024-06-07T06:45:39.291928300Z" level=info msg="Loading containers: done."
```

I concur that this smells like a race condition with WSL2 when using [bind mounts](https://docs.docker.com/storage/bind-mounts/).

**Comment by kanjieater** [comment]
Just want to call out that a side effect to the bind mounts being created too early, seems to resulting in docker not seeing that the folder structures already exists, which if you're using the docker compose shorthand syntax
```
volumes:
	- /path/to/config:/config
```
instead of 
```
    volumes:
      - type: bind
        source: /path/to/config
        target: /config
```
Docker will _always_ create the path prior to the container as root, which causes all kinds of havoc to your permissions if you're trying to run as the usual UID:GID, 1000:1000.

**Comment by longzheng** [comment]
After hours of trial and error testing, I have narrowed down this problem specifically to just bind mounting **files** (not folders) and using **WSL2** for docker-compose.

### Reproduction

The simplest repro I came up with was basically a `docker-compose.yml` with just one service and an individual file mount.

```yml
services:
  influxdb:
    image: influxdb:1.8
    restart: always
    ports:
      - "8086:8086"
    volumes:
      - "./influxdb.conf:/influxdb.conf"
      # alternatively can use the type: bind syntax which has the same issue
      # - type: bind
      #   source: ./influxdb.conf
      #   target: /influxdb.conf
```

Then just create a blank called `influxdb.conf`.

The **most important step to reproduce this issue** is to **initialize the docker container using WSL**. So first type `wsl` to open WSL2 at the current directory, and then `docker compose up -d`.

This will create the container with a bind mount of `/mnt/c/Users/longz/Desktop/test/influxdb.conf`.

![Screenshot 2024-06-08 174712](https://github.com/docker/for-win/assets/484912/275a5ea9-62e0-4fb5-8a82-2a3c7ef7e0e9)

Then if you **log out** or restart and **log back in**, the container will have a status of `Excited (127)` when Docker Desktop is configured to "**start when you sign in to your computer**".

![Screenshot 2024-06-08 174956](https://github.com/docker/for-win/assets/484912/6cedba7a-5cc9-4fc3-be70-8b25337471d9)

### Workaround

I found as a workaround, if you start the docker compose from Windows (command prompt), not WSL2, then it will work correctly. 

For example from the same directory just run `cmd` and then `docker compose up -d`, it will create a container with a bind mount of `C:\Users\longz\Desktop\test\influxdb.conf`. 

![Screenshot 2024-06-08 175325](https://github.com/docker/for-win/assets/484912/0ad18821-a18e-4aec-bc39-00080b3d7551)

This container will start correctly if you log out and log back in.

### Logs

When it fails with WSL bind mount

```
time="2024-06-08T07:46:10.132502900Z" level=info msg="Starting up"
time="2024-06-08T07:46:10.202880400Z" level=info msg="[graphdriver] using prior storage driver: overlay2"
time="2024-06-08T07:46:10.264388100Z" level=info msg="Loading containers: start."
time="2024-06-08T07:46:10.373628100Z" level=info msg="Removing stale sandbox ee9fba78558b36cde4ca5affff4fcca2d893a88ffbfb8429427aefd28ef3497d (f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c)"
time="2024-06-08T07:46:10.390831000Z" level=warning msg="Error (Unable to complete atomic operation, key modified) deleting object [endpoint 9c75cf86e2840d2e6be23069440acb96a53bbe33a624a326d4fab1d18a8c12a7 e3cd2cbe49631a54b53dd8c2aecf07581fee5cf7cb55f59447dcd649e7d28234], retrying...."
time="2024-06-08T07:46:10.435127800Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2024-06-08T07:46:10.594770200Z" level=error msg="stream copy error: reading from a closed fifo"
time="2024-06-08T07:46:10.594825700Z" level=error msg="stream copy error: reading from a closed fifo"
time="2024-06-08T07:46:10.612752400Z" level=error msg="f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c cleanup: failed to delete container from containerd: container \"f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c\" in namespace \"moby\": not found"
time="2024-06-08T07:46:10.806178200Z" level=error msg="failed to start container" container=f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c error="failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/c2760049fc8d5840d923e884d0d5c597983b14a35a07dd23e3fce9b8f080fc27\" to rootfs at \"/influxdb.conf\": mount /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu/c2760049fc8d5840d923e884d0d5c597983b14a35a07dd23e3fce9b8f080fc27:/influxdb.conf (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type"
time="2024-06-08T07:46:10.806268600Z" level=info msg="Loading containers: done."
time="2024-06-08T07:46:10.817468500Z" level=warning msg="WARNING: No blkio throttle.read_bps_device support"
time="2024-06-08T07:46:10.817501300Z" level=warning msg="WARNING: No blkio throttle.write_bps_device support"
time="2024-06-08T07:46:10.817508100Z" level=warning msg="WARNING: No blkio throttle.read_iops_device support"
time="2024-06-08T07:46:10.817511200Z" level=warning msg="WARNING: No blkio throttle.write_iops_device support"
time="2024-06-08T07:46:10.817513800Z" level=warning msg="WARNING: daemon is not using the default seccomp profile"
time="2024-06-08T07:46:10.817526600Z" level=info msg="Docker daemon" commit=de5c9cf containerd-snapshotter=false storage-driver=overlay2 version=26.1.4
time="2024-06-08T07:46:10.817779000Z" level=info msg="Daemon has completed initialization"
time="2024-06-08T07:46:10.922223200Z" level=info msg="API listen on /var/run/docker.sock"
time="2024-06-08T07:46:10.922306400Z" level=info msg="API listen on /run/guest-services/docker.sock"
time="2024-06-08T07:46:11.884586300Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=9734f53b5c010f77 traceID=70a85b04dae6b8e1eed61279492a76ab
time="2024-06-08T07:51:22.402821015Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=809fa9480d55520e traceID=98f750795508c8218c910caedc28f803
time="2024-06-08T07:51:22.406775154Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=e430ad3d6c1f2dc4 traceID=a5b8a96e770a274da06ceb5a6e91faa6
time="2024-06-08T07:51:22.411005403Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=2e503f3c80cd35fc traceID=491ad4e2ea75f2c13a88bbd50a1afdcf
time="2024-06-08T07:51:22.414956343Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=2801fdf650d385ba traceID=51d82f952b6d1b1a41351a7aeac544e1
time="2024-06-08T07:51:22.418894181Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=07566d24fad9a43e traceID=fff263ebb8faa89bc8359defc738243b
time="2024-06-08T07:51:22.423088929Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=63e06060e803f29a traceID=ce1a9e8048ebaaf80e66637012c8e4de
time="2024-06-08T07:51:22.426915364Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=67f6a0ea5cf3ddd5 traceID=b59eeb439bd14e430b3549f9d709fd4a
time="2024-06-08T07:51:22.431097212Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=5375960a670549ff traceID=8cd07ea66a2ae145aa20c4d16c24e335
time="2024-06-08T07:51:22.435321061Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=0915dc967b22d00d traceID=65527c08ea4ef391692c0a055c860f5b
time="2024-06-08T07:51:22.439222499Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=7cbfc67b3b663796 traceID=b08a0b3ec66a9c1fac2aa9a685455749
time="2024-06-08T07:51:45.128346791Z" level=info msg="ignoring event" container=f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2024-06-08T07:51:45.133030458Z" level=warning msg="ShouldRestart failed, container will not be restarted" container=f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c daemonShuttingDown=false error="restart canceled" execDuration=46.080874533s exitStatus="{0 2024-06-08 07:51:45.123416215 +0000 UTC}" hasBeenManuallyStopped=true restartCount=0
time="2024-06-08T07:51:45.441325975Z" level=error msg="Error setting up exec command in container f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c: container f1d46c06e9dd76be4cb6684d52bd57836656ae9ec97198999dde453829ff555c is not running" spanID=42a5f1be7da194e9 traceID=2177681882773035bfc111be4e726530
time="2024-06-08T07:52:13.687265926Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=0e7a632f8f142946 traceID=33e1215940e158c1c3dde3d350cf7e37
```

When it works with Windows bind mount

```
time="2024-06-08T07:53:02.925570900Z" level=info msg="Starting up"
time="2024-06-08T07:53:02.988704100Z" level=info msg="[graphdriver] using prior storage driver: overlay2"
time="2024-06-08T07:53:03.055067900Z" level=info msg="Loading containers: start."
time="2024-06-08T07:53:03.157277400Z" level=info msg="Removing stale sandbox 3d85c11f9c934d488f10d1faaa333eac7ef6426463253ad3c79252bfb40f1e78 (e50abc924a28470c95bfda22bbc72d7957eda01e3fa43a3a156bf9a869b819c8)"
time="2024-06-08T07:53:03.174143800Z" level=warning msg="Error (Unable to complete atomic operation, key modified) deleting object [endpoint 9c75cf86e2840d2e6be23069440acb96a53bbe33a624a326d4fab1d18a8c12a7 c24396f04d3dc71a80e5f6b27dc8057000cc6d00445a2d492815297daffb0028], retrying...."
time="2024-06-08T07:53:03.221468600Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2024-06-08T07:53:03.681588800Z" level=info msg="Loading containers: done."
time="2024-06-08T07:53:03.694146000Z" level=warning msg="WARNING: No blkio throttle.read_bps_device support"
time="2024-06-08T07:53:03.694185400Z" level=warning msg="WARNING: No blkio throttle.write_bps_device support"
time="2024-06-08T07:53:03.694194000Z" level=warning msg="WARNING: No blkio throttle.read_iops_device support"
time="2024-06-08T07:53:03.694198900Z" level=warning msg="WARNING: No blkio throttle.write_iops_device support"
time="2024-06-08T07:53:03.694203000Z" level=warning msg="WARNING: daemon is not using the default seccomp profile"
time="2024-06-08T07:53:03.694220200Z" level=info msg="Docker daemon" commit=de5c9cf containerd-snapshotter=false storage-driver=overlay2 version=26.1.4
time="2024-06-08T07:53:03.694458700Z" level=info msg="Daemon has completed initialization"
time="2024-06-08T07:53:03.843328700Z" level=info msg="API listen on /var/run/docker.sock"
time="2024-06-08T07:53:03.843348300Z" level=info msg="API listen on /run/guest-services/docker.sock"
time="2024-06-08T07:53:05.140474500Z" level=error msg="Handler for GET /v1.45/services returned error: This node is not a swarm manager. Use \"docker swarm init\" or \"docker swarm join\" to connect this node to swarm and try again." spanID=0cd989d415ce81bc traceID=4203602522acced71f2b593de56cf083
```

**Comment by does1tmatter** [comment]
> I tried downgrading to [`4.24.2`](https://docs.docker.com/desktop/release-notes/#4242) and and it works. So the bug is between that version and `4.28`. Don't really feel like testing all the version inbetween, but feel free to chime in if any later version works.

Tried 4.26.0 seems to be working fine. So between 4.26 - 4.28 (just wanted to try first 2024 release at least for the time being)

**Comment by kasprk** [comment]
Running 4.35.1, still the same issue

**Comment by InsightForesight** [comment]
4.36.0 + wsl2
some container couldn't  mount volume after computer reboot, but another container would works fine


```

(HTTP code 500) server error - error while mounting volume '/var/lib/docker/volumes/ollama-intel-gpu_ollama-intel-gpu/_data': failed to mount local volume: mount /run/desktop/mnt/host/wsl/docker-desktop-bind-mounts/Ubuntu-24.04/74ac2895c9940b8efaa0d7647d785ba915437b567f6537a73f6aa08b22d4a789:/var/lib/docker/volumes/ollama-intel-gpu_ollama-intel-gpu/_data, flags: 0x1000: no such file or directory

```



```
# docker-compose-wsl2.conf

  ollama-intel-gpu: 
    driver: local
    driver_opts:
      type: none
      device: /home/volumes/ollama/intel-gpu
      o: bind
```

**Comment by kanjieater** [comment]
Can anyone confirm if this is still an issue on 2025-01-30's 4.38.0? 
[4.38.0](https://docs.docker.com/desktop/release-notes/#4380)

I want to try getting nvidia gpu running inside wsl2, and people are recommending updating to later versions, but I don't want to run into this.

**Comment by singhaditya4** [comment]
> Can anyone confirm if this is still an issue on 2025-01-30's 4.38.0? [4.38.0](https://docs.docker.com/desktop/release-notes/#4380)
> 
> I want to try getting nvidia gpu running inside wsl2, and people are recommending updating to later versions, but I don't want to run into this.

still an issue after it restarts with sudden power loss

**Comment by dario-innatial** [comment]
Still an issue with 4.43.2.

Evey day I start my computer and have to manually restart the containers in order for the volumes to work.

**Comment by kanjieater** [comment]
v4.44.2 still has this issue and 4.27.2, the last version that didn't have this issue wont work with cuda & GPU enabled docker containers


Race conditions related to volumes also happens on a pure ubuntu system running docker, so  you have to I worked around that by making systemd depend on
```[Unit]
After=zfs-mount.service
Requires=zfs-mount.service
```

Unfortunatley on windows, the issue is docker starts and docker says the host volumes exist, and tries to load them but they pointing to a different file system that has root (since the real one hasn't loaded yet). The work around, is to disable Docker Desktop from starting automatically with windows. Then you can use a powershell script like this, and have it run in your startup directory (minimized). All this does is make sure wsl2 has started w/ systemd, and when it does, start docker desktop after that.
```
# ===================================================================================
# wsl2Startup.ps1 (v3 - Text Output Check)
# Description: This script waits for the WSL systemd service to be fully running
#              by checking the text output of the 'systemctl' command, working
#              around a bug where wsl.exe reports an incorrect exit code.
# ===================================================================================

Write-Host "Starting WSL2 readiness check..."

# Loop indefinitely until the systemd state is 'running' or 'degraded'.
while ($true) {
    # Execute the WSL command and capture its text output into the $wsl_state variable.
    $wsl_state = wsl.exe -d Ubuntu -u ke --exec bash -c "sudo systemctl is-system-running"

    # Trim whitespace from the output and check if it's one of the success states.
    if ($wsl_state.Trim() -eq 'degraded' -or $wsl_state.Trim() -eq 'running') {
        # If it is, break out of the infinite loop.
        break
    }
    
    # If not in a success state, print the current state and wait before retrying.
    Write-Host "WSL systemd state is '$wsl_state'. Not ready yet. Retrying in 10 seconds..."
    Start-Sleep -Seconds 2
}

# Once the loop exits, it means the command returned a success state.
Write-Host "Success! WSL systemd state is '$wsl_state'. Launching Docker Desktop..."

# Start the Docker Desktop process.
Start-Process "C:\Program Files\Docker\Docker\Docker Desktop.exe"

Write-Host "Script has finished."
```


This is my hacky work around until fix this.

**Comment by helltraitor** [comment]
Same issue on `Version 4.86.0 (236216)`, `WSL2`, `Windows 11 Pro (25H2) x86_64`, `WIN32_NT 10.0.26200.9168`

Maybe you can add at least an option to delay some containers (or entire stacks for first time, for simplicity) to wait on wsl disks?

I was very wondered when I  unable to login into my own vaultwarden (dyndns setup).

So I tried to fix it with "hypervisor" container.

<details>
<summary>If someone will find this helpfull</summary>

Barely tested (since bug is rare). But seems to be work.
Note: after start will recreate stacks (you can replace down & up with just up if needed)

```
.
├── compose.yml
└── build
    └── service
        ├── Dockerfile
        ├── entrypoint.sh
        └── utils
            └── log.sh
```

`compose.yml`

```yaml
name: supervisor

services:
  service:
    build: ./build/service
    image: supervisor:latest
    container_name: supervisor-main
    network_mode: none
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
```

`build/service/Dockerfile`

```dockerfile
FROM docker:29.7.2-cli-alpine3.24

RUN apk add --no-cache bash jq

COPY entrypoint.sh /entrypoint.sh
COPY utils /utils

RUN chmod +x /entrypoint.sh

ENTRYPOINT ["/entrypoint.sh"]
```

`build/service/entrypoint.sh`

```bash
#!/bin/bash


printf '\033[0;90m[NOTE   ]\033[0m: ==================== CONTAINER START ====================\n'
printf '\033[0;90m[NOTE   ]\033[0m: Begin loading utils\n'

for file in /utils/*; do
  source "$file"
  printf '\033[0;90m[EXEC   ]\033[0m:     %s ok\n' "$(basename "$file")"
done

###############################################################################
##                          Gather stack information                         ##
###############################################################################

log "Gather stack information"

stacks=$(
  docker ps -aq --filter label=com.docker.compose.project |
  xargs -r docker inspect |
  jq '
    group_by(.Config.Labels["com.docker.compose.project"])
    | map({
        key: .[0].Config.Labels["com.docker.compose.project"],
        value: {
          mounts: [
            .[].Mounts[]
            | select(.Type == "bind")
            | .Source
            | capture("^(?<mount>/mnt/[^/]+)")?
            | .mount
          ] | unique,
          compose: .[0].Config.Labels["com.docker.compose.project.config_files"]
        }
      })
    | map(select(.value.mounts | length > 0))
    | from_entries
  '
)

log "Found compose stacks:"
log.from jq -r 'keys[]' <<< "$stacks"

###############################################################################
##                        Begin check uncovered mounts                       ##
###############################################################################

log "Begin check uncovered mounts"

uncovered=$(
  jq -r '
    to_entries[]
    | . as $stack
    | [
        $stack.value.mounts[] as $mount
        | select(($stack.value.compose | startswith($mount + "/")) | not)
        | $mount
      ] as $mounts
    | select($mounts | length > 0)
    | "\($stack.key) have uncovered mounts: \($mounts | join(", "))"
  ' <<< "$stacks"
)

if [ -n "$uncovered" ]; then
  LOG_FROM__LEVEL=warn log.from printf '%s\n' "$uncovered"
  log "Required script update" warn
fi

unset uncovered

###############################################################################
##                           Begin check used mounts                         ##
###############################################################################

log "Begin check used mounts"

checks=$(
  jq -r '
    to_entries[]
    | .value as $stack
    | $stack.mounts[] as $mount
    | select($stack.compose | startswith($mount + "/"))
    | "\($mount)|\($stack.compose)"
  ' <<< "$stacks" |
  sort -t'|' -k1,1 -u
)

until (
  for check in $checks; do
    mount="${check%%|*}"
    compose_file="${check#*|}"
    relative_path="${compose_file#"$mount"/}"

    log.from docker run --rm \
      --mount "type=bind,src=$mount,dst=/check,readonly" \
      alpine:3.24.1 \
      test -f "/check/$relative_path" || {
        log "Mount check failed: $mount" note
        exit 1
      }
  done
); do
  log "Check unsuccessful, retrying after 1 second" note
  sleep 1
done

unset checks

###############################################################################
##                      Begin compose stacks recreation                      ##
###############################################################################

log "Begin compose stacks recreation"

for stack in $(jq -r 'keys[]' <<< "$stacks"); do
  compose_file=$(jq -r --arg stack "$stack" '.[$stack].compose' <<< "$stacks")
  stack_path=$(dirname "$compose_file")

  log "Recreate stack: $stack"

  if log.from docker run --rm \
    -v /var/run/docker.sock:/var/run/docker.sock \
    --mount "type=bind,src=$stack_path,dst=$stack_path" \
    -w "$stack_path" \
    docker:29.7.2-cli-alpine3.24 \
    sh -c '
      docker compose down
      docker compose up -d
    '
  then
    log "Stack recreated: $stack" success
  else
    log "Stack recreation incomplete: $stack" error
  fi
done

log "All actions applied, see you soon!"
sleep infinity
```

`build/service/utils/log.sh`

```bash
#!/bin/bash

log.from() {
  local level="${LOG_FROM__LEVEL:-exec}"

  "$@" 2>&1 | while read -r line; do
    log "    $line" "$level"
  done

  return "${PIPESTATUS[0]}"
}

function log() {
  COLOR_GRAY="\e[0;90m"
  COLOR_CYAN="\e[0;36m"
  COLOR_YELLOW="\e[0;33m"
  COLOR_RED="\e[0;31m"
  COLOR_GREEN="\e[0;32m"
  #
  COLOR_RESET="\e[0m"
  #
  COLOR_CURRENT=""
  LEVEL=""

  case $2 in
    # COMMON LEVELS
    err*)
      COLOR_CURRENT=$COLOR_RED
      LEVEL="[ERROR  ]"
      ;;
    warn*)
      COLOR_CURRENT=$COLOR_YELLOW
      LEVEL="[WARNING]"
      ;;
    debug)
      COLOR_CURRENT=$COLOR_GRAY
      LEVEL="[DEBUG  ]"
      ;;
    # SPECIFIC
    note)
      COLOR_CURRENT=$COLOR_GRAY
      LEVEL="[NOTE   ]"
      ;;
    exec)
      COLOR_CURRENT=$COLOR_GRAY
      LEVEL="[EXEC   ]"
      ;;
    success)
      COLOR_CURRENT=$COLOR_GREEN
      LEVEL="[SUCCESS]"
      ;;
    # DEFAULT
    *)
      COLOR_CURRENT=$COLOR_CYAN
      LEVEL="[INFO   ]"
      ;;
  esac

  echo -e $LOG__ECHO_FLAGS "${COLOR_CURRENT}${LEVEL}${COLOR_RESET}: $1"
}
```
</details>

**Comment by helltraitor** [comment]
Sorry got same on this solution (because of socket, maybe tcp needed)
